### DISCUSSION TITLE
Where conditions for class members and properties

### DISCUSSION THREAD

**Discussion by osamakawish** [dotnet/csharplang#5472] [Language Ideas] [answered]
As the title says, is it possible to produce results where I can use a where condition on a class member's or a property's type. The main reason for this is that I'm using a class defined by someone else, so I can't just edit it to implement an interface.

For example, I want to be able to do something like the following:

```c#
public interface ICanOkay 
{
    public Button OkButton { get; }
}

public class C 
{
    public TUIElement CanOkayElement { get; } where TUIElement : UIElement, ICanOkay
}
```

Even a workaround would be fair, as long as it's not too big.

## Edit: Possible Solution

I think a simply solution would be a using alias:

```c#
using HasOkayButtonElement = ICanOkay UIElement;

public class C : HasOkayButtonElement
{
    public HasOkayButtonElement CanOkayElement { get; }
}
```

## Edit: Additional Details

(Copied from comments)

Say, you've got 2 classes:

```c#
public class A : UIElement, ICanOkay { ... }
public class B : UIElement, ICanOkay { ... }
```
The point is, that the variable `CanOkayElement` accepts instances of both `A` and `B`.

I'm not specializing a template, so much as referring to a "category" of types. And it accepts instances of any object that implements the specified parent class and interface, since `UIElement` itself does not implement `ICanOkay`.

It's worth noting that this can definitely be converted to a property as follows:

```c#
public void SetCanOkayElement<TUIElement>(TUIElement element) where TUIElement : UIElement, ICanOkay { ... }
public TUIElement GetCanOkayElement<TUIElement>() where TUIElement : UIElement, ICanOkay { ... }
```

My point is that it doesn't make sense to require casting here when storing it as a variable. Nor should I need to rely on reflection. I want to let the compiler know that this is a variable that accepts instances of `UIElement` if they're of a class that implements `ICanOkay`. It allows me to access the `OkayButton` method of `ICanOkay` during compile time, as well as the remaining methods of `UIElement`.

### Repeated `where` statements
To add, this cleans up a lot of of repeated `where` statements, especially that are repeated, and makes it easier to refactor, especially when similar where conditions are reused repeatedly.

```c#
using HasOkayButtonElement = ICanOkay UIElement;

public class MyMessageBox : HasOkayButtonElement { ... }
public class MyOpenFileDialog : HasOkayButtonElement { ... }
```

### Minimizes need for abstract classes

Quite simply, I no longer need to create an entire abstract class like the following that I repeatedly implement:

```c#
public abstract class AbstractHasOkayButtonElement : UIElement, ICanOkay { ... }
```

The code block in the section for repeated where statements simplifies this already.

### Refactoring Benefits

In situations where certain inheritances are reused, only one statement really needs updating (though the code below also updates the class name, which is possible with existing refactoring methods in VS anyways):

```c#
public interface ICanCancel { CancelButton { get; } }

using HasOkayCancelButtonsElement = ICanOkay ICanCancel UIElement;

public class MyMessageBox : HasOkayCancelButtonsElement { ... }
public class MyOpenFileDialog : HasOkayCancelButtonsElement { ... }
```

As opposed to multiple ones:

```c#
public interface ICanClose { CloseButton { get; } }

public class MyMessageBox : UIElement, ICanClose, ICanOkay { ... }
public class MyOpenFileDialog : UIElement, ICanClose, ICanOkay { ... }
```

This reasoning can likely be applied to numerous collection/enumerable interfaces that we already have in .Net.

# Summary

It cleans up code, improves refactoring, and has the additional benefit of accessing methods from multiple interfaces. And as far I'm aware, this isn't conflicting with existing where condition to break any existing code.

Pretty sure this covers the usual criticisms I've received in the past for my ideas lol.

**Comment by huoyaoyuan** [comment]
This is effectively generic specialization, which is not supported by CLR yet.

**Reply by osamakawish** [reply]
I'm not sure I see how this is generic specialization... The `HasOkayButtonElement` is sort of a reference to any object that's an instance of a class inheriting `UIElement` and `ICanOkay`. It's not exactly a specific implementation of a generic. I can use template parameters or repeatedly cast to work around this, but it should be more efficient to simply recognize that it's an instance of a class implementing `ICanOkay` that inherits `UIElement`. Needless to say, it's also much more clearly expressed that that's what this is.

I'm not making a specialized implementation of a class at all.

I think you're misunderstanding the implementation (mostly because I didn't clearly cover that), but here it is.

Say, you've got 2 classes:

```c#
public class A : UIElement, ICanOkay { ... }
public class B : UIElement, ICanOkay { ... }
```

The point is, that the variable `CanOkayElement` accepts instances of both `A` and `B`.

I'm not specializing a template, so much as referring to a "category" of types. And it accepts instances of any object that implements the specified parent class and interface, since `UIElement` itself does not implement `ICanOkay`.

It's worth noting that this can definitely be converted to a property as follows:

```c#
public void SetCanOkayElement<TUIElement>(TUIElement element) where TUIElement : UIElement, ICanOkay { ... }
public TUIElement GetCanOkayElement<TUIElement>() where TUIElement : UIElement, ICanOkay { ... }
```

My point is that it doesn't make sense to require casting here when storing it as a variable. Nor should I need to rely on reflection. I want to let the compiler know that this is a variable that accepts instances of `UIElement` **_if_** they're of a class that implements `ICanOkay`. It allows me to access the `OkayButton` method of `ICanOkay` during compile time, as well as the remaining methods of `UIElement`.

**Comment by osamakawish** [comment]
If I don't get any more objections, I'm likely going to turn this into a proper proposal.

**Reply by CyrusNajmabadi** [reply]
How would you implement this?

**Reply by osamakawish** [reply]
I don't really understand the code behind of the language, but I'd reckon that however the objects store which class an object is an instance of, it would store all the interfaces + class for the given variable that are contained in the using condition. We can call these things "type groups".

And once the type group itself (ie. the using statement stating the multiple types) is modified, it modifies all references to have said updated implementations.

It depends on how the code-behind already works regarding how types of objects are identified.

It would update variables in much the same way that we already update classes by adding a new interface to implement to an existing class in c# right now.

Which part specifically are you asking of the implementation of? I assumed you meant how VS would raise warnings/errors based on modifications of the code here. But I may have been wrong in assuming that.

**Reply by osamakawish** [reply]
> How would you implement this?

If you mean how I would implement it for members/properties, I've already demonstrated what methods the properties would get converted into in this example:

```c#
public void SetCanOkayElement<TUIElement>(TUIElement element) where TUIElement : UIElement, ICanOkay { ... }
public TUIElement GetCanOkayElement<TUIElement>() where TUIElement : UIElement, ICanOkay { ... }
```

This should be the same as:

```c#
using HasOkayButtonElement = ICanOkay UIElement;

public HasOkayButtonElement CanOkayElement { get; set; }
```

The hard part is class members that aren't properties. But for that, it's likely similar to having classes/interfaces with where conditions. And again, this may be me misunderstanding something, but it really depends on what specifically you think would be difficult to implement.

**Reply by osamakawish** [reply]
I can demonstrate how I'd do this right now, it's just that I need to repeatedly cast to `ICanOkay` just to access the `OkayButton`. This should be accessible at compile time. Again, it's just letting the compiler know that the variable should be of a class that implements the specified interfaces and class.

So any variable declaration like this:
```c#
SomeType variableName
```

Doesn't strictly always test the type of the variable during assignment, but if `SomeType` was part of a using alias, it checks it the class of the variable `variableName` implements all the interfaces and the class (if any) of the using condition.

This, for example:
```c#
using X = Interface1 Interface2 A;
X x;
```
tests whether or not the class for x is of type A, implements Interface1 and Interface2, during its assignment. Currently, it just needs to check if the type given to variable `x` is of type `X` during assignment. But now, it just has to check a little bit more when we're using these "grouped types" (ie. the using statement `using X = Interface1 Interface2 A;`).

**Reply by osamakawish** [reply]
If still confused, let me know what and where. My explanation may have been vague.

**Comment by TheUnlocked** [comment]
Probably related: #4628/#399.

**Reply by osamakawish** [reply]
Very related. My code only considers "Intersection types" and groups them under a using alias. You can toss union types and group them just as well.

In fact, once those are implemented, I think this would be just a natural use of how using aliases already work, in combination with union/intersection types.

**Comment by osamakawish** [comment, marked as answer]
Proposal published over at #5476.
